Create a Resource When A Client Submits an HTML Form

omarnyte opened the pull request:
## Main
In order to direct POST requests toward the proper `Handler`, I've added the following route to the `routes.put("/api/form", new FormHandler(store));`.

## RequestParser 
Although this class previously extracted an incoming `Request`'s status line and passed it to `StatusLine` for parsing, it was not doing so correctly. I previously stringified the incoming request's stream inside the `ClienThread` class so that I could pass in that result to `RequestParser`. I would pass that string as an argument to `RequestParser`, which would attempt to extract the status line by splitting on `\r\n`. Naively, the process of stringifying the request stream using a `BufferedReader` removed any line separations.

Because those line separations are vital for parsing the request, I decided to pass the stream as an argument to `RequestParser`. This move allowed me to continue using a `BufferedReader`'s `.readLine()` to simplify parsing the status line and the headers. Because the `messageBody` is comprised of key value pairs separated by `&`, I only need to use the more primitive `.read()` method when parsing the message body.  

## FormHandler
I've updated the `FormHandler` since the IPM. Instead of creating a resource that relies on specific keys in the `messageBody`, the class now creates a simple formatted text file that echoes the `Request`'s `messageBody`. 

## DataStore
In order to enable different `DataStore`s to create files in any way their specific implementations require, I've added the `#postFile(String uri, byte[] content)` method to the `DataStore` interface.

## Directory
The `Directory` class creates a resource via the `#postFile(String uri, byte[] content)` method. It uses it's `directoryPath` (which is set at its construction and chosen by the user when starting the server) and appends the URI of the file that will be created.

Commit:
Create POSTable-form.html

Commit:
Integrate Headers and MessageBody as HashMaps in Request class

Commit:
Configure RequestParer to take a BufferedReader as a constructor arg

Moving the logic for reading from the BufferedReader allows the
RequestParser to parse the status line, headers, and messsageBody

Commit:
Add 201 Created, 303 See Other, and 400 Bad Request

Commit:
Create a new text file with a randomly generated name

FormHandler takes the submitted form data and prints the data to a text
file

Commit:
Move file posting logic to DataStore / Directory

Commit:
Assert that response generated by FormHandler correctly sets reasonPhrase and Location Header field

Commit:
Create a txt resource based on the data POSTed from POSTable-form.html

Commit:
Change ResponseHeader and ResponseHeaderTest to MessageHeader and MessageHeaderTest

jayden commented on src/main/java/Request/RequestParser.java:
```diff
@@ -1,30 +1,73 @@
+      stringBuilder.append((char) (this.reader.read()));
     }
+    
+    return new String(stringBuilder);
```
This can just be `stringBuilder.toString();`

jayden commented on src/main/java/Request/RequestParser.java:
```diff
@@ -1,30 +1,73 @@
-      this.requestLine = new RequestLine(requestLineString);
-    } catch (Exception e) {
-      throw new BadRequestException("Bad request!");
+  private Request.Builder buildRequestLine(Request.Builder requestBuilder) throws IOException {
```
You aren't the return values of any of these private methods here. Why not just return each value and build the request in `generateRequest`?

```java
RequestLine requestLine = new RequestLine(this.read.readLine());
HashMap<String, String> headers = parseHeaders();
HashMap<String, String> body = parseMessageBody();

return new Request.Builder()
  .method(requestLine.getMethod())
  .uri(requestLine.getURI())
  .version(requestLine.getHTTPVersion())
  .headers(headers)
  .body(body)
  .build();
```

omarnyte replied:
I've followed this suggestion, which actually allowed me to improve my exception handling. Instead of catching `ArrayIndexOutOfBoundException` and `IOException` inside the `#generateRequest` method, I know catch those exceptions inside every individual helper method. Upon catching them, I throw a `BadRequestException` whose message points to the part of the incoming request that caused the exception.  This approach allows `ClientThread` to catch the `BadRequestException` and build the correct `Response` for the end user while simultaneously providing useful information to clients of the class.

jayden reviewed (commented):
I also cannot get your integration tests to pass so I cannot build the JAR. Can you create a custom gradle task to run the integration tests and separate them from `gradle build`?

omarnyte commented on src/main/java/Handler/FormHandler.java:
```diff
@@ -0,0 +1,51 @@
+  }
+
+   private byte[] createFileContent(Request request) {
+    String LINE_SEPARATOR = System.getProperty("line.separator");
```
The variable was leftover from the previous way I was handling POST requests in this class. Since I was previously responding to HTML, I forgot to change the `Content-Type: text/html` header, so the entire message body was appearing as one line. I've removed the variable now.

Commit:
Set a Request's headers and messageBody by passing in a HashMap

Commit:
Throw BadRequestException from RequestLine when passed in bad request line

Commit:
Improve RequestParser exception handling

By moving exception catching to private methods, thrown
BadRequestExceptions can include more descriptive error messages

omarnyte commented on src/main/java/Request/RequestParser.java:
```diff
@@ -1,30 +1,73 @@
+      buildMessageBody(requestBuilder);
+      return requestBuilder.build();
+    } catch (ArrayIndexOutOfBoundsException e) {
+      throw new BadRequestException("Bad request from ArrayIndex!");
```
See below.
